lambda/translcore.ml

```diff
@@ -592,51 +603,91 @@ and transl_exp0 ~in_new_scope ~scopes e =
         Llet(Variable, Pintval, counter, (int 0),  
           Lsequence(transl_for arr block_count, Lvar(res))))
     in
-    let gloabl_counter = Ident.create_local "gloabl_counter" in
     let block_count = List.length blocks in 
+    let global_counter = 
+      if (block_count = 1)
+      then None
+      else Some(Ident.create_local "global_counter") 
+    in
     let base_block = 
-      transl_base_block (List.hd blocks) gloabl_counter (block_count = 1) 
+      transl_base_block (List.hd blocks).clauses global_counter 
     in
-    (*TODO: We Should be able to get rid of this variable anb just use the 
+    (*TODO: We Should be able to get rid of this variable and just use the 
       most toplevel one created by transl_block. *)
     let arr = Ident.create_local "arr" in 
     let translated_blocks = 
-      List.fold_left transl_block base_block (List.tl blocks) 
+      List.fold_left transl_block base_block 
+        (List.map (fun a -> a.clauses) (List.tl blocks))
     in
-    if block_count = 1 then 
-      (*No need to concat if there is only the base block.*)
-      translated_blocks
-    else
-      Llet(
-        Variable, Pintval, gloabl_counter, int 0,
-        Llet(
-          Strict, Pgenval, arr,
-          translated_blocks,
-          transl_concat_arrays (Lvar(arr)) block_count gloabl_counter
-        ))
+    (match global_counter with
+        | None -> translated_blocks
+        | Some gc -> 
+          Llet(
+            Variable, Pintval, gc, int 0,
+            Llet(
+              Strict, Pgenval, arr,
+              translated_blocks,
+              transl_concat_arrays (Lvar(arr)) block_count gc
+            )))
   | Texp_list_comprehension (body, blocks) -> 
-    (*The left most statement should be at the end of the recursion.*)
-    let block = List.concat blocks in
-    let transl_list_comp type_comp bdy is_concat = 
-      let return = 
-        if is_concat 
-        then Pgenval 
-        else Typeopt.value_kind body.exp_env body.exp_type 
+    let int n = Lconst (Const_base (Const_int n)) in  
+    let block = List.fold_left (fun acc block -> 
+        let rec loop = function 
+          | [] -> assert false 
+          | [last] -> [last, block.guard]
+          | l::ls -> (l, None) :: (loop ls) 
+        in
+        List.concat [List.rev (loop block.clauses); acc]
+        ) [] blocks
+    in
+    let rec transl_list_comp block upper_res= 
```

Comment:
This function is very messy and hard to follow. I think that most of the complexity will disappear by the observation that there is no need for separate `concat` and `prepend` versions of the primitives, or any need for different handling of the `single_loop` case. However, I think it would be useful to think about and discuss how you might implement this function more clearly.